csrc/python_frontend/segmentation.cpp

```diff
@@ -79,6 +79,167 @@ int64_t SegmentationState::setupSegmentation(
+
+  std::vector<int64_t> original_fid;
+  original_fid.reserve(cloned_inputs.size() + cloned_outputs.size());
+
+  std::transform(
+      cloned_inputs.begin(),
+      cloned_inputs.end(),
+      std::back_inserter(original_fid),
+      [&](Val* v) { return map_cloned_value_to_fid_.at(v); });
+
+  std::transform(
+      cloned_outputs.begin(),
+      cloned_outputs.end(),
+      std::back_inserter(original_fid),
+      [&](Val* v) { return map_cloned_value_to_fid_.at(v); });
+
+  // Step 4b) ir_cloner maps cloned fusion Vals to segment Vals.
+  std::vector<Val*> segment_inputs_outputs;
+  segment_inputs_outputs.reserve(cloned_inputs.size() + cloned_outputs.size());
+
+  std::transform(
+      cloned_inputs.begin(),
+      cloned_inputs.end(),
+      std::back_inserter(segment_inputs_outputs),
+      [&](Val* v) { return cloned_to_segment_map.clone(v); });
+
+  std::transform(
+      cloned_outputs.begin(),
+      cloned_outputs.end(),
+      std::back_inserter(segment_inputs_outputs),
+      [&](Val* v) { return cloned_to_segment_map.clone(v); });
+
+  // Step 4c) Map segment Vals to their FusionDefinition index.
+  std::vector<int64_t> segment_fid;
+  segment_fid.reserve(segment_inputs_outputs.size());
+  std::transform(
+      segment_inputs_outputs.begin(),
+      segment_inputs_outputs.end(),
+      std::back_inserter(segment_fid),
+      [&](Val* v) { return map_translated_val_to_segment_fid.at(v); });
+
+  // Step 4d) Map original indices to segment indices.
+  NVF_ERROR(original_fid.size() == segment_fid.size());
+  std::unordered_map<int64_t, int64_t> segment_fid_to_original_fid_map;
+  for (size_t idx : c10::irange(original_fid.size())) {
+    segment_fid_to_original_fid_map.emplace(
+        segment_fid.at(idx), original_fid.at(idx));
+  }
+
+  // Step 4e) short-circuit: No extra extents required for python definition.
+  if (fusion_segment->inputs().size() == other.inputs().size()) {
+    return segment_fid_to_original_fid_map;
+  }
+
+  // The python segment can require the size of tensor dimensions from original
+  // fusion's input arguments, which the CPP segment does not.
+
+  // Step 4f) Create a map from segment to cloned extents.
+  // Step 4g) Create a map from segment indices to segment extents.
+  std::unordered_map<Val*, Val*> segment_to_cloned_extents;
+  std::unordered_map<size_t, Val*> segment_fid_to_translated_val;
+  for (Val* cloned_extent : cloned_extents_) {
+    Val* segment_extent = cloned_to_segment_map.clone(cloned_extent);
+
+    // short-circuit: some extents are not used in segment
+    if (map_translated_val_to_segment_fid.count(segment_extent) == 0) {
+      continue;
+    }
+
+    size_t segment_fid = map_translated_val_to_segment_fid.at(segment_extent);
+    segment_to_cloned_extents.emplace(segment_extent, cloned_extent);
+    segment_fid_to_translated_val.emplace(segment_fid, segment_extent);
+  }
+
+  // Step 4h) Find the set difference between all segment input indices and
```

Comment:
I see the point of having the two map above.
But from step 4h) -> 4l), we are just doing a chained mapping. Do we really want to have it done in such an explicit way? other than just do the mapping in

```
  for (int64_t input_fid : other.inputs()) {
    if (segment_fid_to_original_fid_map.count(input_fid) == 0) {
      segment_fid_to_original_fid_map.emplace(input_fid, magic_chained_map(input_fid));
    }
  }
```

Reply:
The `magic_chained_map` can represented as a function to do the following:
`segment index -> segment val -> cloned val -> original index`